### ISSUE TITLE
UI changes needs to be communicated to the doc team

### ISSUE THREAD

**Issue by rhuss** [syndesisio/syndesis-ui#1097]
Update / Changes in the UI need to be actively communicated to the doc team

**Comment by gashcrumb** [comment]
@TovaCohen what would be your preferred workflow for this?  I guess also UXD may need to also be involved, WDYT @dongniwang?

**Comment by dongniwang** [comment]
Right now, for new features, we're trying to have internal ux review then team review, doc is invited to join the team review. But for existing enhancements, I think maybe tracking ui and ux issues with proper labels?

**Comment by TovaCohen** [comment]
Is there a way to just select a checkbox when you commit or merge or push a UI update?  I am not clear on the exact point where such a checkbox would be useful. 
The reason I suggested this is because there were a number of changes that I missed. For example, OAuth Client Authorization is now OAuth Application Authorization. I also missed that in the Settings page, when you insert values for a particular app, you have to click Save and then OK rather than just OK. And there was another button label change that I missed. I found out about these changes because Fintan was kind enough to capture them by following the tutorial instructions while actually using TP1. So the tutorial instructions were wrong in quite a few places, even though at each standup I asked for help making sure that the instructions were correct.  Maybe I am doing something wrong and there is already a way for me to know about these kinds of changes. Maybe I need a lesson on how to use this interface. The way it is now, I get too much information so maybe I need to filter out notifications that have no UI impact. 
Thanks very much for following up!!

**Comment by TovaCohen** [comment]
I should also say that I did not actually use TP1 to follow my own instructions. I had a deadline on another project (Fuse tooling). But I am going through the tutorial steps right now while actually doing everything. The thing is, I cannot do this every day in order to pick up every single update.

**Comment by gashcrumb** [comment]
I would say that really any given PR could carry significant UI changes with it, we could mention interested parties in PRs easily enough, though I'd be concerned that'd add to the noise.  And if it becomes every PR or issue, you may as well just watch the whole repo :-)

Maybe need to be better about adding documentation tasks along with UX, UI and backend tasks to the epics?

Other option is we could also tag issues that require UI changes and add a `documentation` tag perhaps, and then the issue can only be closed once there's a PR for code changes *and* a related documentation update?

**Comment by amysueg** [comment]
Just reading though this and had some thoughts.

In response to @dongniwang and @TovaCohen 's comments, it seems that sometimes changes are made to the product UI without either UX or documentation knowing about them. Applying a UX label when changes are pushed will help draw attention to those changes while they are happening. But we'd like to know about UI changes (whether based on designs we provide or changes developers introduce into the UI) before they are pushed to ensure consistency. I may not be following this completely, but it seems that we all want to be better about this so I'd be happy to talk through it.

**Comment by TovaCohen** [comment]
I hesitate to say this, but I'm going to say it anyway. For me, for user doc, TP1 is not history. I'm actively trying to make sure the doc for TP1 is correct and complete. I am saying this because I feel that I have failed at integrating the delivery of the doc with the delivery of the product. And I don't know how to fix that. 

There is a syndesis-documentation repo and a bunch of weeks ago, Fintan created the doc issues we thought were on the horizon. That didn't work out to be helpful. 

I've followed the discussion of monorepo a bit, but I have been afraid to ask how this affects user doc because I didn't think I would understand the answer.   I'm sorry to be such a dummy about all this. 

In summary, I don't know how to leverage this GitHub user interface to deliver the user doc at the same time, or as soon as possible tafter,  you deliver the product.

**Comment by amysueg** [comment]
Well @TovaCohen I certainly don't think you're a dummy! You are asking all the right questions, IMHO!

**Comment by dongniwang** [comment]
@TovaCohen I think the whole process needs to be improved! And agreed with @amysueg that you're asking all the right questions. I think the Monorepo move affects us all. And we definitely need a plan for it.

**Comment by rhuss** [comment]
@TovaCohen all of your questions and concerns are valid. Thanks for raising them !

We are not so good in pinging other teams when there are relevant updates. I also think, that a "push" model (where the developer will actively notify other teams) will never work, because its often not on the developer's radar and can be forgotten quickly who to notify. Also, as mentioned above, the automatic notifications by GitHub is not useful if it exceeds a certain threshold.

So I recommend a "pull" model, where we teams which depend on be notified of code changes (including UI updates) periodically. We can use labels for this. For documentation, I recommend introducing a label "documentation change". The issues with this can be easily queried with the ZenHub board (by an URL which can be bookmarked). So when such an issues pop up, @TovaCohen will see this issue on the board, can update the documentation and then _remove_ the "documentation label" so that it disappears from this view again.

The question now is, who should attach this "documentation change" label ?  First of all, this should be the developer working on the issue / PR. But this again is something which can be easily forgotten. However, I think this a good first step (if we raise awareness for this, too). In the next step, we can automate this and e.g., automatically apply this label if a file from a specific directory has been touched by a PR (e.g. if an HTML has been changed). This heuristic will not be perfect but should cover 90% of all cases.

wdyt ? @TovaCohen, if you think this is a good idea, we can set up a Bluejeans call and I would go with you through the steps of this workflow.

**Comment by TovaCohen** [comment]
Thanks @rhuss !
I think I understand your suggestion. I am not so sure of the mechanics of actually doing it. 
Maybe a BJ call would be the best way to sort this out. 

I want to clarify that I am not trying to address the need for new documentation for a new feature in the UI. That's a separate process and deserves attention of its own. I am also not trying to address changes where UX and doc should be involved in the discussion of the change before it is implemented. 

I want to clarify my narrow scope because of Stan's comment:

> "I would say that really any given PR could carry significant UI changes with it, we could mention interested parties in PRs easily enough, though I'd be concerned that'd add to the noise. And if it becomes every PR or issue, you may as well just watch the whole repo :-)"

Stan, does "the whole repo" apply to the syndesis-ui repo?

The particular need I am trying to address is an update to a (typically small) part of the UI that is already available and that has already been documented. For example, the label on a button changed from "Deploy" to "Publish". To meet that need, would it be possible for a new "**doc update needed**" label to be applied to a PR by the person who merges the PR immediately before or after merging the PR? Then I know that it is done and I can update the doc, add a comment to the PR to say the doc was updated , and remove the label.

Correct me if I am wrong, but with your suggestion Roland, I think I would need to watch for comments in issues and PRs that have the "**doc update required**" label to find out if and when the planned update is merged. 

To see what has the label, in the board at 

https://github.com/syndesisio/syndesis-project#boards

I would click the Labels button and select the new "**doc update needed**" label as the filter, right?

If we do create a new label, I'd like it to be "**doc update needed**". 
Who creates the new label?

We already have the "**support and docs**" label -- what is that used for?

**Comment by zregvart** [comment]
One idea might be to validate the docs by making them executable tests, at least a part of them. I've discussed this before but perhaps it would merit to create a workshop or brainstorm this in detail: we could include a BDD scenario as a part of feature and in these make very explicit links between UI and text of the scenario.
This way it could be seen explicitly if the text of the UI changed. For UX it might be beneficial to have end to end tests record either videos or screenshots and use a visual comparison tool to highlight changes and require review.
This will require investment in tooling but I think it will pay off in the long run.

**Comment by kahboom** [comment]
@TovaCohen - There is no such thing as a "dummy" question. 😉  That label is supposed to be for pages, for instance, where attention is required or assistance is needed from the docs team. We have a similar one for UXD, but since you can't really subscribe to labels out of the box (more on that later), we end up having to notify UXD via chat anyway and give them a heads up about the open issues, but since they have labels so they are easier to keep track of. It definitely doesn't solve the problem.

@zregvart - E2E tests could potentially be a good approach, but it will be specific to changes that affect the docs and UI teams. I think we should be doing this anyway though; it's good practice to do screenshots of E2E tests and fail them if what the user expects is not displayed.

I think following entire repos has taken a toll on my productivity levels, and it's wayyy too much noise. GitHub email notifications are going to be the death of me. If I unsubscribe from the API repo then I'll never get notified unless I'm specifically tagged. This creates an inherent problem for cross-collaborating across teams.

I've thought about this a lot, but can really only think of two things that help to solve this without getting too complicated:

1. **Label subscriptions & properly using GH teams**. I think maybe adding more labels is sometimes convenient but can become less useful if we need to be maintaining them and can't subscribe to them. They should be used to help us quickly filter things. An idea here is that we write a script to subscribe to them via CD/CI. For instance, use the GitHub webhooks for PR or issues events and check when specific labels are added and send notifications to respective teams (e.g. "uxd" to the UXD team): https://developer.github.com/v3/activity/events/types/#issuesevent

2. **GH PR templates & properly using GH teams**. Another idea is that whenever there is a pull request opened, regardless of which repository it's from (hopefully a single one in the future), we could require that a specific GitHub template be used for PRs, as is traditionally done in open source community projects, even for GH issues. This will encourage consistency across the development process, similarly to how Commitizen attempts to standardize GH commits with a specific format. For instance, the template could have a "Screenshots", and "Potential Affected Areas" section, which will prompt the collaborator (whether it be someone from UXD, docs, QE, UI, etc), to consider the areas that would be affected and then instead of tagging individuals, tagging with @syndesisio/design-ux for instance, if they wanted to inform UXD.

3. A combination/variation of both of the above. I think if the progress of the issue is being blocked, then a Blocked column could be more helpful here, as the columns are meant to illustrate the progress of a particular task, whereas labels are meant for easier filtering and identification. If the progress is impacted due to Docs or UXD intervention being required, that should be clear from the board. If they just need to be notified, but it isn't blocking, at that point a single notification should suffice and it doesn't matter if the label is kept or removed (up to them). Also, would all of this just be for PRs or issues as well?

We could go as far as to suggest labels based on the content that is changed, but that would be very difficult to do and likely, given a lower accuracy, not worth the effort. There is seemingly no way around requiring manual identification of what areas could potentially be affected and which teams should be notified. This is mostly human intuition imo.

@zregvart - Btw, do we have some type of proposal doc for the monorepo movement that could perhaps educate the rest of the team about how we can leverage model- or feature-driven planning and see everything at a glance? I'd love to help contribute once I have a better grasp on how it could help improve the work flow. Another use case is how we can better use the Zen Hub board with a monorepo.

**Comment by rhuss** [comment]
@TovaCohen you described the suggested workflow very accurately. I create the label but would like to call it sth like "UI changed" (so from the POV of the source, not the target which can be more than only documentation updates).

@zregvart I completely agree with you that having testable UI content is a very good thing if even only a failing test reminds to apply a label (which again could be in theory be done automatically). 

@kahboom:

* You can subscribe to GH issues of course, but you can't filter notifications based on labels. But I think we can easily do this outside of GH, too (via purebot or a custom hubot). Labels are the canonical way to organize issues and implement workflows, there is no alternative in the GitHub world. I opened https://github.com/syndesisio/syndesis/issues/9 for discussion about a more concise and consistent label structure.
* Templates are a good idea, but since they are 'free' form they are often ignored, too. It's good for bug reports to collect as many information as required. For driving a workflow is not appropriate as it's in the text to an issue which is not visible when looking at lists. Labels are much better for this.
* I agree that we can and should rethink the way how we use ZenHub board when going monorepo.

**Comment by rhuss** [comment]
I suggest the following approach (in this order):

* We start immediately with the label based approach as this is the easiest one. Every PR which affects the text content of the UI should be labeled 'text changed' (other suggestions for the labels ?). This is a quick win.
* We we will look into purebot how we can apply this label automatically (e.g. when an HTML page has been changed)
* We evaluate how to enable Hubot to notify automatically via chat when this label has been applied (which mimics the current approach we UX related 'pings' are transported over IRC, too)
* We are looking into the big solution by connecting documentation creation with the UI and automated testing.

**Comment by zregvart** [comment]
> @zregvart - Btw, do we have some type of proposal doc for the monorepo movement that could perhaps educate the rest of the team about how we can leverage model- or feature-driven planning and see everything at a glance? I'd love to help contribute once I have a better grasp on how it could help improve the work flow. Another use case is how we can better use the Zen Hub board with a monorepo.

I'm not aware of one 😿. I think we could brainstorm a bit about how this would work and add this to our engineering guidelines. I have some experience with BDD and I have an idea of how this might work and the QE team has already a lot of this in place. It requires big investment in tooling that I think pays off.

**Comment by kahboom** [comment]
@rhuss:

> For driving a workflow is not appropriate as it's in the text to an issue which is not visible when looking at lists. Labels are much better for this.

I agree with you on this. I probably misunderstood; I thought the goal was to notify respective teams when there's a relevant change (since you can't subscribe to all issues for a particular label), without having to subscribe to an entire repo, but not that it was a problem that required changing the existing workflow.

Side note: I think it would also help if labels could have descriptions, but that's out of our hands I guess.

**Comment by TovaCohen** [comment]
Thanks everyone for the thoughtful responses. 
Sure, let's give the "text changed" label a try. Is this how it will work:
- UI engineer applies the "text changed" label to a PR that contains code that changes the text that the user sees in the UI
- Doc team filters PRs to monitor those with the "text changed" label.
- When a PR with the "text changed" label is merged, doc team updates the doc

Sounds like there is no need to remove the label. Correct me if I have anything wrong. 
But if I understand correctly, then this issue can be closed. I'm a happy camper again :-)

**Comment by rhuss** [comment]
> Sounds like there is no need to remove the label. Correct me if I have anything wrong.

Removing the labels helps you to track the items which still needs to be done. But if its ok for you that these tasks might still show up when you filter on this, then we can leave the label on the issue, too. It's up to you ;-)

**Comment by rhuss** [comment]
Just created the label. So using https://github.com/syndesisio/syndesis-project#boards?labels=text%20changed&repos=89959255,96111131,63097460,75404146,92834472,82667482,96436677,92716590,93627906,81928401,95247826,96449862 you can easily see the PR's tagged with this label now.

**Comment by TovaCohen** [comment]
Hmmmm.  I would remove the label after the PR is merged and after I update the doc.
But I don't want to remove it if anyone else needs it. 
@dongniwang @sjcox-rh @amysueg  - do you anticipate needing to see PRs with this label after the PR has been merged?

**Comment by rhuss** [comment]
We could also automatically add a new label, only for this purpose and which is completely owned by you. E.g when someone sets the label "text changed" then purebot automatically can add an extra label "documentation update required", which you are free to delete (the "text changed" label still would stay).

**Comment by TovaCohen** [comment]
Sure. That would be good. Then I can filter on "doc update required", and remove it when I've done the work. Thanks!

**Comment by dongniwang** [comment]
> Hmmmm. I would remove the label after the PR is merged and after I update the doc.
But I don't want to remove it if anyone else needs it.
@dongniwang @sjcox-rh @amysueg - do you anticipate needing to see PRs with this label after the PR has been merged?

I think it's fine that you remove the label after merging the PR. 

I'm just thinking out loud here, is there a way to add some kind of check list to the process, for changes in UI, we need UX review and doc review?

**Comment by rhuss** [comment]
I think thats a good reason to keep the 'text changed' label, as there might be two teams which are interested in those changes. So, in order to provide a checklist I would just add two extra labels autoamatically: "doc update required", "ux review required".

**Comment by TovaCohen** [comment]
Sounds good to me.  Thanks!

**Comment by TovaCohen** [comment]
I am not seeing any of these labels on the ZenHub board:
text changed
doc update required
ux review required

I did previously see the "text changed" label and was able to filter on it. 
This is the link I use to display the board:
https://github.com/syndesisio/syndesis-project#boards?notFullScreen=false&repos=92834472
I am viewing all repositories. When I click the down caret next to Label, none of those labels are appearing. Any suggestions for me?

**Comment by rhuss** [comment]
@TovaCohen as we have moved now all open issues to syndesisio/syndesis, we introduced there a label "notif/doc-update" for this purpose. Currently there is no issues assigned. @syndesisio/ui-api please take care in the future if there are UI label changes. Should also be part of the review.

thanks ..
